libraries/chain_kv/include/b1/session/session.hpp

```diff
@@ -842,92 +929,300 @@ const std::shared_ptr<const cache_data_store>& session<persistent_data_store, ca
+        // This can happen because there are various levels of sessions with possibly duplicate keys.
+        // Just move to the next key and try again.
+        // If the key at the top is coming from a session where that key has been deleted in a child session
+        // Just move to the next key and try again.
+        if (is_deleted_(top_key, index_(*top)) || top_key == previous_key) {
+          move_top();
+          continue;
+        }
+
+        // We have a new current key.  Set it and return.
+        m_current = top;
+        return;
     }
 }
 
 template <typename persistent_data_store, typename cache_data_store>
 template <typename iterator_traits>
 void session<persistent_data_store, cache_data_store>::session_iterator<iterator_traits>::move_next_() {
     // increment the current iterator and set the next current.
-    auto predicate = [](auto& it) { 
+    auto move = [](auto& it) { 
         if (it.begin == it.end) {
             // There is nothing to iterate over.
             return;
         }
 
-        if (it.current == it.end || ++it.current == it.end) {
-            // Loop around to the beginning.
-            it.current = it.begin;
+        if (it.current == it.end) {
+            // This lambda doesn't perform the roll over.
+            return;
         }
+
+        ++it.current;
     };
-    move_(predicate, std::less<bytes>{});
+
+    auto rollover = [](auto& it) {
+      if (it.begin == it.end) {
+          // There is nothing to iterate over.
+          return;
+      }
+      it.current = it.begin;
+    };
+
+    // Prime the queue.
+    auto top = m_next.top();
+    m_next.pop();
+    m_next.push(top);
+
+    move_(m_next, move, rollover);
 }
 
 template <typename persistent_data_store, typename cache_data_store>
 template <typename iterator_traits>
 void session<persistent_data_store, cache_data_store>::session_iterator<iterator_traits>::move_previous_() {
     // decrement the current iterator and set the next current.
-    auto predicate = [](auto& it) { 
+    auto move = [](auto& it) { 
         if (it.begin == it.end) {
             // There is nothing to iterate over.
             return;
         }
 
         if (it.current == it.begin) {
-            // Loop around to the end.
+            // set to the end.
             it.current = it.end;
+            return;
         }
 
         --it.current;
     };
 
-    move_(predicate, std::greater<bytes>{});
-}
+    auto rollover = [](auto& it) {
+      if (it.begin == it.end) {
+          // There is nothing to iterate over.
+          return;
+      }
+      it.current = it.end;
+      --it.current;
+    };
 
-template <typename persistent_data_store, typename cache_data_store, typename iterator_traits>
-using session_iterator_alias = typename session<persistent_data_store, cache_data_store>::template session_iterator<iterator_traits>;
+    // Prime the queue.
+    auto top = m_previous.top();
+    m_previous.pop();
+    m_previous.push(top);
```

Comment:
I'm not sure I understand what this was supposed to do. Seems like a no-op.

Reply:
The problem is, for example, that we could be incrementing along and while this is updating the queue that is holding the value of the next iterator, the queue that is holding the values for the previous container is not updating because it nothing is being pushed/popped to it.  So the idea was to force an update on the queue before using just to make sure it was correctly sorted before asking for the top element.